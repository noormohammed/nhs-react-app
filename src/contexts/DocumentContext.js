import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useRef,
} from "react";

/**
 * DocumentContext for Document details
 */

const DocumentContext = createContext({});

/**
 * Document Provider using the DocumentContext
 * @param {*} children any props/children components passed to the DocumentContextProvider
 * @returns memoized object of document info/details state, set state function
 */
const DocumentContextProvider = ({ children }) => {
  // Once the document info is memoized, we store the info in the localStorage
  // for ease of access and its also economical (can avoid too many backend calls).
  const [documentInfo, setDocumentInfo] = useState(
    window.localStorage.getItem("document_info")
      ? JSON.parse(window.localStorage.getItem("document_info"))
      : null
  );

  const isMountedRef = useRef(null);

  useEffect(() => {
    isMountedRef.current = true;
    const storeDocumentInfo = async () => {
      if (isMountedRef.current) {
        window.localStorage.setItem("document_info", JSON.stringify(documentInfo));
      }
    };
    storeDocumentInfo();

    return () => {
      isMountedRef.current = false;
    };
  }, [documentInfo]);

  return (
    <DocumentContext.Provider value={[documentInfo, setDocumentInfo]}>
      {children}
    </DocumentContext.Provider>
  );
};

const useDocumentContext = () => useContext(DocumentContext);

export { useDocumentContext, DocumentContextProvider };
