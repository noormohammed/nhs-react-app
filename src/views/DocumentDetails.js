import React, { useEffect } from "react";
import { useHistory } from "react-router-dom";

import { useDocumentContext } from "contexts/DocumentContext";
import Page from "components/Page";
import DocumentTypeOptions from "static/document_types";
import DocumentUsers from "static/document_users";
import SampleDocumentDetails from "static/sample_document_details";

// Using the modes defined in the environmental variables
const { REACT_APP_DOC_READ_MODE, REACT_APP_DOC_EDIT_MODE } = process.env;

const DocumentSection = ({ labelClass, label, testId, children }) => {
  return (
    <div className="row">
      <div className={labelClass}>{label}</div>
      <div className="column" style={{ padding: "20px 10px" }} data-testid={testId}>
        {children}
      </div>
    </div>
  );
};

const DocumentDetails = ({ mode }) => {
  const history = useHistory();
  const [documentInfo, setDocumentInfo] = useDocumentContext();

  useEffect(() => {
    // Document details can be fetched / pulled here from an API
    // For the sake this test, we are fetching from a json file
    const getDocumentDetails = () => {
      if (!documentInfo || Object.entries(documentInfo).length <= 1) {
        setDocumentInfo({
          ...SampleDocumentDetails,
          mode: mode,
        });
      } else {
        setDocumentInfo({
          ...documentInfo,
          mode: mode,
        });
      }
    };

    getDocumentDetails();
    return () => {};
  }, []);

  // Extracting name from array object
  const findNameFromArray = (givenArray, condition) => {
    return givenArray
      .filter((obj) => {
        return obj.id === condition;
      })
      .map((item) => item.name)[0];
  };

  const handleClickForEditField = (e, fieldAttributes) => {
    e.preventDefault();
    setDocumentInfo({
      ...documentInfo,
      mode: "edit-field",
    });
    history?.push({
      pathname: "/edit-field",
      state: fieldAttributes,
    });
  };

  return (
    <Page
      title={mode === REACT_APP_DOC_EDIT_MODE ? "Edit Document" : "Document Details"}
    >
      <div className="document">
        {documentInfo?.mode === REACT_APP_DOC_READ_MODE ? (
          <div className="heading">DOCUMENT INFORMATION</div>
        ) : (
          <div className="heading">Make additions to the document below.</div>
        )}
      </div>
      <div className="table">
        <div className="table-body">
          <DocumentSection
            labelClass={`label column-heading column  ${
              mode === REACT_APP_DOC_EDIT_MODE && "required"
            }`}
            label="NAME"
            testId="name"
          >
            <button
              className="edit-detail-button"
              onClick={(e) =>
                mode === REACT_APP_DOC_EDIT_MODE &&
                handleClickForEditField(e, {
                  title: "Enter Name",
                  type: "textarea",
                  required: "required",
                  name: "name",
                  maxLength: 50,
                  rows: 1,
                  cols: 50,
                })
              }
            >
              {documentInfo?.name}
              {mode === REACT_APP_DOC_EDIT_MODE && (
                <span className="caret right"></span>
              )}
            </button>
          </DocumentSection>

          <DocumentSection
            labelClass={`label column column-heading ${
              mode === REACT_APP_DOC_EDIT_MODE && "required"
            }`}
            label="TYPE"
            testId="type"
          >
            <button
              className="edit-detail-button"
              onClick={(e) =>
                mode === REACT_APP_DOC_EDIT_MODE &&
                handleClickForEditField(e, {
                  title: "Select Type",
                  type: "select",
                  required: "required",
                  name: "type",
                  options: DocumentTypeOptions,
                })
              }
            >
              {findNameFromArray(
                DocumentTypeOptions,
                parseInt(documentInfo?.type, 10)
              )}
              {mode === REACT_APP_DOC_EDIT_MODE && (
                <span className="caret right"></span>
              )}
            </button>
          </DocumentSection>

          <DocumentSection
            labelClass="label column column-heading"
            label="DESCRIPTION"
            testId="description"
          >
            <button
              className="edit-detail-button"
              onClick={(e) =>
                mode === REACT_APP_DOC_EDIT_MODE &&
                handleClickForEditField(e, {
                  title: "Enter Description",
                  type: "textarea",
                  name: "description",
                  maxLength: 100, //This value can be increased for a real time app
                  rows: 3,
                  cols: 50,
                })
              }
              style={{ fontSize: 18 }}
            >
              {documentInfo?.description}
              {mode === REACT_APP_DOC_EDIT_MODE && (
                <span className="caret right"></span>
              )}
            </button>
          </DocumentSection>

          <DocumentSection
            labelClass="label column column-heading"
            label="USERS ALLOWED"
            testId="allowed_users"
          >
            <button
              className="edit-detail-button"
              onClick={(e) =>
                mode === REACT_APP_DOC_EDIT_MODE &&
                handleClickForEditField(e, {
                  title: "Select Users Allowed Access",
                  type: "select",
                  multiple: true,
                  name: "allowed_users",
                  options: DocumentUsers,
                })
              }
            >
              {documentInfo?.allowed_users?.map((item, index) => {
                let userName = findNameFromArray(DocumentUsers, parseInt(item, 10));
                return `${userName}${
                  documentInfo.allowed_users.length - 1 === index ? "" : ", "
                }`;
              })}
              {mode === REACT_APP_DOC_EDIT_MODE && (
                <span className="caret right"></span>
              )}
            </button>
          </DocumentSection>
        </div>
      </div>
    </Page>
  );
};

export default DocumentDetails;
