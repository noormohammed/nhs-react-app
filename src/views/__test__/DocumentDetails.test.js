import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createBrowserHistory } from "history";
import { Router } from "react-router-dom";

import { DocumentContextProvider } from "contexts/DocumentContext";
import DocumentDetails from "views/DocumentDetails";

// Using the modes defined in the environmental variables
const { REACT_APP_DOC_READ_MODE, REACT_APP_DOC_EDIT_MODE } = process.env;

const browserHistory = createBrowserHistory();

const documentInfo = null;
const setDocumentInfo = jest.fn();

describe("<DocumentDetails /> in read-only mode", () => {
  let getByTestId;
  beforeEach(() => {
    ({ getByTestId } = render(
      <DocumentContextProvider value={[documentInfo, setDocumentInfo]}>
        <Router history={browserHistory}>
          <DocumentDetails mode={REACT_APP_DOC_READ_MODE} />
        </Router>
      </DocumentContextProvider>
    ));
  });

  afterEach(cleanup);

  test("renders DocumentDetails correctly", () => {
    expect(getByTestId).toMatchSnapshot();
  });

  test("renders correct title", () => {
    expect(screen.getByText(/Document Details/i)).toBeTruthy();
  });

  test("renders the Document Name", () => {
    expect(getByTestId("name")).toHaveTextContent("Sample Document");
  });

  test("renders the Document Type", () => {
    expect(getByTestId("type")).toHaveTextContent("Standard");
  });

  test("renders the Document Description", () => {
    expect(getByTestId("description").textContent).toContain(
      "To be completed by doctors."
    );
  });

  test("renders the Document Users Allowed to access", () => {
    expect(getByTestId("allowed_users").textContent).toContain("User 2, User 3");
    expect(getByTestId("allowed_users").textContent).not.toContain("User 4");
  });

  test("check if Menu button exists", () => {
    expect(getByTestId("menu-button")).toBeTruthy();
    expect(getByTestId("menu-button")).toBeInTheDocument();
    expect(getByTestId("menu-button").textContent).toBe("⚙️");
  });

  test("trigger Menu button click and view changed to Document Menu", () => {
    const menuBtn = screen.getByText("⚙️");
    expect(menuBtn).toBeInTheDocument();

    userEvent.click(menuBtn);
    expect(browserHistory.length).toEqual(2);
    expect(browserHistory.location.pathname).toBe("/document-menu");
  });
});

describe("<DocumentDetails /> in edit mode", () => {
  let container;
  beforeEach(() => {
    ({ container } = render(
      <DocumentContextProvider value={[documentInfo, setDocumentInfo]}>
        <Router history={browserHistory}>
          <DocumentDetails mode={REACT_APP_DOC_EDIT_MODE} />
        </Router>
      </DocumentContextProvider>
    ));
  });

  afterEach(cleanup);

  test("renders DocumentDetails correctly", () => {
    expect(container).toMatchSnapshot();
  });

  test("renders DocumentDetails correctly", () => {
    expect(screen.getByText(/Edit Document/i)).toBeTruthy();
  });

  test("check if document has six buttons", () => {
    const allBtns = screen.getAllByRole("button");
    expect(allBtns).toHaveLength(6);
  });

  test("check if document has four carets icons for edit buttons", () => {
    expect(container.getElementsByClassName("caret").length).toBe(4);
  });

  test("check if document has four edit buttons", () => {
    const allEditBtns = container.getElementsByClassName("edit-detail-button");
    expect(allEditBtns.length).toBe(4);
  });

  test("trigger Document Detail edit click to push edit field view", () => {
    const allEditBtns = container.getElementsByClassName("edit-detail-button");

    userEvent.click(allEditBtns[0]);
    expect(browserHistory.location.pathname).toBe("/edit-field");
  });

  test("trigger Tick ✅ button click to return back to Document Details view", () => {
    const tickBtn = screen.getByTestId("edit-done");

    userEvent.click(tickBtn);
    expect(browserHistory.location.pathname).toBe("/");
  });
});
