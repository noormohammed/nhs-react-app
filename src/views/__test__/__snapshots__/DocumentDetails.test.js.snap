// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocumentDetails /> in edit mode renders DocumentDetails correctly 1`] = `
<div>
  <div
    class="main"
  >
    <div
      style="position: fixed; top: 0px; display: flex; flex-direction: row; justify-content: center; align-items: center; width: 100%; height: 50px; background-color: rgb(63, 80, 181); border-bottom: 1px solid #d8d8d8; font-weight: bold; padding: 0px 20px 0px 0px; box-sizing: border-box; color: rgb(255, 255, 255); font-size: 18px;"
    >
      <div
        style="width: 50px;"
      >
        <button>
          ⬅️
        </button>
      </div>
      <div
        style="margin-left: auto; margin-right: auto;"
      >
        Edit Document
      </div>
      <div
        style="width: 50px;"
      >
        <button
          data-testid="edit-done"
        >
          <span>
            ✅
          </span>
        </button>
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="document"
      >
        <div
          class="heading"
        >
          Make additions to the document below.
        </div>
      </div>
      <div
        class="table"
      >
        <div
          class="table-body"
        >
          <div
            class="row"
          >
            <div
              class="label column-heading column  required"
            >
              NAME
            </div>
            <div
              class="column"
              data-testid="name"
              style="padding: 20px 10px;"
            >
              <button
                class="edit-detail-button"
              >
                Sample Document
                <span
                  class="caret right"
                />
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="label column column-heading required"
            >
              TYPE
            </div>
            <div
              class="column"
              data-testid="type"
              style="padding: 20px 10px;"
            >
              <button
                class="edit-detail-button"
              >
                Standard
                <span
                  class="caret right"
                />
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="label column column-heading"
            >
              DESCRIPTION
            </div>
            <div
              class="column"
              data-testid="description"
              style="padding: 20px 10px;"
            >
              <button
                class="edit-detail-button"
                style="font-size: 18px;"
              >
                To be included as part of the Adult Admission Package. To be completed by doctors.
                <span
                  class="caret right"
                />
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="label column column-heading"
            >
              USERS ALLOWED
            </div>
            <div
              class="column"
              data-testid="allowed_users"
              style="padding: 20px 10px;"
            >
              <button
                class="edit-detail-button"
              >
                User 1, 
                User 2, 
                User 3
                <span
                  class="caret right"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; align-items: center; justify-content: center; width: 100%; height: 40px; background-color: rgb(51, 51, 51); color: rgb(255, 255, 255); position: fixed; bottom: 0px;"
    >
      Footer Menu
    </div>
  </div>
</div>
`;

exports[`<DocumentDetails /> in read-only mode renders DocumentDetails correctly 1`] = `[Function]`;
